### DISCUSSION TITLE
Override rule configuration in an "imported" ruleset

### DISCUSSION THREAD

**Discussion by ddidier** [pmd/pmd#3517] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**

I have defined a "baseline" _ruleset_ file `pmd/pmd-base.xml` e.g.

```xml
<ruleset name="Base ruleset" ...>
  ...
  <rule ref="category/java/codestyle.xml">
    <exclude name="LocalVariableCouldBeFinal" />
    <exclude name="MethodArgumentCouldBeFinal" />
  </rule>
  <rule ref="category/java/codestyle.xml/LongVariable">
    <properties>
      <property name="minimum" value="25" />
    </properties>
  </rule>
  ...
</ruleset>
```

This one is packaged in a common Maven artifact.

Other Maven projects depends on this artifact and have their own _ruleset_ file `pmd/pmd-custom.xml` e.g.

```xml
<ruleset name="Custom ruleset" ...>
  <rule ref="pmd/pmd-base.xml">

     <!-- this is working -->
    <exclude name="LinguisticNaming" />

     <!-- this is not working -->
    <rule ref="category/java/codestyle.xml/LinguisticNaming">
    <!-- or <rule ref="LinguisticNaming"> -->
      <properties>
        <property name="booleanMethodPrefixes" value="something else" />
      </properties>
    </rule>

     <!-- this is not working -->
    <rule ref="category/java/codestyle.xml/LinguisticNaming">
    <!-- or <rule ref="LinguisticNaming"> -->
      <properties>
        <property name="violationSuppressRegex" value=".*" />
      </properties>
    </rule>

  </rule>
  ...
</ruleset>
```

**Describe the solution you'd like**

I don't want to pollute my code with `@SuppressWarnings`.

I'd like to override only parts of a custom _ruleset_ as above.

**Describe alternatives you've considered**

Having only a custom _ruleset_ by project but this is not DRY.

**Additional context**

N/A

**Comment by oowekyala** [comment, marked as answer]
You can just move the rule configuration outside of the ruleset import element:
```xml
<ruleset name="Custom ruleset">
  <rule ref="base.xml">
    <!-- Exclude the base configuration -->
    <exclude name="LinguisticNaming" />
  </rule>

  <!-- Specify your own configuration -->
  <rule ref="category/java/codestyle.xml/LinguisticNaming">
    <properties>
      <property name="booleanMethodPrefixes" value="something else" />
    </properties>
  </rule>
</ruleset>
```

If you're ok with this solution I'll turn this issue into a discussion to improve its discoverability

**Comment by ddidier** [comment]
That's working, thanks!

**Comment by adangel** [comment]
Just for completeness - nesting a rule element inside a rule element is not allowed by our schema (https://pmd.sourceforge.io/ruleset_2_0_0.xsd). We seem to ignore such invalid elements, but future versions of PMD might reject the whole ruleset, if it doesn't validate against the schema.

**Comment by rsoesemann** [comment]
@oowekyala or @adangel is it also possible to reference a ruleset from a link (e.g. the Bitbucket location of our Company ruleset).

I would want all our companies repos and projects to reference our default rulset without having to copy it to their repo root.
